### PR TITLE
Move to grpc-dotnet

### PR DESCRIPTION
**Pull request by xlegalles**
It targets net5.0;net6.0;net7.0 but I don't know if net5.0 is still relevant (not supported any more). 
Also, tests only target net7.0.
I propose to update this package to v2.0 but of course, you decide.

I have tried to remove as many warnings as possible.

All unit tests now use a standard host instead of a Grpc.Core.Server which does not exist anymore. In that context, GatewayTestService cannot be resolved from DI. This is why I have created IRequestHandlerRegistration which encapsulate the logic for tests: it's a singleton component whose state can be reset.
In BaseZeebeTest, I have changed the lifecycle of this host and made it unique for all tests: the execution of tests is of course much faster.

The tricky part was about certificates and I've put some comments in ZeebeSecureClientBuilder.

One important remark in ZeebeClient: because the certificates we use for tests are untrusted, I have to add a RemoteCertificateValidationCallback line 86 to bypass validation. I'm not sure if this is relevant? In our case, we often use OpenSSL certificates and we have to use such a validation. But at a consequence the ShouldFailOnWrongCert UT does not work anymore.
 
I'm available for any discussion/comments but maybe not before 21th as we have public holidays in France.

### CODE REVIEW

**Comment by xlegalles** [comment]
resolves https://github.com/camunda-community-hub/zeebe-client-csharp/issues/348

**Review by Zelldon — COMMENTED** [review]
Hey @xlegalles first thanks that you did this and provided a solution for it. Right now I'm have not really time to maintain this project, we try to find someone else to take over that. Furthermore, I was off the last few weeks (dependabot gets merged automatically), so sorry for the late response.


I haven't looked at all details in your PR, but what I have seen is that there are a lot of changes that change code style and move classes/files. This makes it quite hard to review (it is not clear what are the important parts and what not).. Over 1k lines of changes are not optimal to review in general. Would be nice if we could avoid the style changes. If we need them let's either create a separate commit for it or better create a separate PR for it.

Furthermore, please follow the [contribution guide](https://github.com/camunda-community-hub/zeebe-client-csharp/blob/main/CONTRIBUTING.md), and adjust your commit messages etc.

File: Client.UnitTests/ZeebeClientTest.cs
Comment (xlegalles) [inline Client.UnitTests/ZeebeClientTest.cs:1]: This test class has been splitted because some tests require a test server while others do not need any.

File: Client.UnitTests/CancelProcessInstanceTest.cs [hunk 1]
```diff
@@ -18,6 +19,7 @@ public async Task ShouldSendRequestAsExpected()
             {
                 ProcessInstanceKey = 12113
             };
+            TestService.Reset();
```
Comment (Zelldon) [inline Client.UnitTests/CancelProcessInstanceTest.cs:22]: :question: For what these resets have been added?

File: Client/Client.csproj [hunk 2]
```diff
@@ -125,16 +126,13 @@ This release is based on the Zeebe 8.1.X release (https://github.com/zeebe-io/ze
...
-    <PackageReference Include="Grpc.Auth" Version="2.54.0" />
-    <PackageReference Include="Grpc.Core" version="2.46.6" />
+    <PackageReference Include="Grpc.Net.Client" version="2.54.0" />
+    <PackageReference Include="Grpc.AspNetCore" version="2.54.0" />
+    <PackageReference Include="Grpc.Auth" version="2.54.0" />
     <PackageReference Include="Grpc.Tools" version="2.55.1">
       <PrivateAssets>all</PrivateAssets>
       <IncludeAssets>runtime; build; native; contentfiles; analyzers</IncludeAssets>
     </PackageReference>
-    <PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="7.0.0" />
```
Comment (Zelldon) [inline Client/Client.csproj:136]: :question: There have been some dependencies removed. How does this work now?

File: Client/Impl/Builder/ZeebeClientBuilder.cs [hunk 3]
```diff
@@ -111,19 +117,50 @@ public IZeebeClientFinalBuildStep UseAccessTokenSupplier(IAccessTokenSupplier su
...
             return this;
         }
 
         public IZeebeClient Build()
         {
-            return new ZeebeClient(Address, Credentials, keepAlive, sleepDurationProvider, loggerFactory);
+            return new ZeebeClient(Address, Credentials, KeepAlive, SleepDurationProvider, LoggerFactory, Certificate);
+        }
+
+        private X509Certificate2 CreateFromPemFile(string fileName)
```
Comment (Zelldon) [inline Client/Impl/Builder/ZeebeClientBuilder.cs:135]: :x: Feels like this shouldn't be part of this PR, is this another feature you want to get merged?

File: Client/Impl/CustomHeaderInterceptor.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+using Grpc.Core.Interceptors;
+
+namespace Zeebe.Client.Impl
+{
+    /// <summary>
+    ///     Intercept outgoing call to inject metadata
+    ///     The "user-agent" is already filled by grpc-dotnet
+    ///     typically something like: key=user-agent, value=grpc-dotnet/2.53.0 (.NET 7.0.5; CLR 7.0.5; net7.0; windows; x64)
+    /// </summary>
+    internal class CustomHeaderInterceptor : Interceptor
```
Comment (Zelldon) [inline Client/Impl/CustomHeaderInterceptor.cs:11]: :question: Was this extracted? Why the change of the user agent?

**Review by Zelldon — COMMENTED** [review]
Hey @xlegalles 

sorry that it took me so long. I'm quite busy with other stuff (zeebe etc.)

I have some questions which we should clarify before we continue I think, see below.

I used the emoji code for the review, which we use also @ Zeebe. For more information take a look here https://github.com/camunda/zeebe/blob/main/CONTRIBUTING.md#reviewing-a-pull-request

File: Client/Impl/Builder/ZeebeClientBuilder.cs (hunk 3 above)
Comment (Zelldon) [reply Client/Impl/Builder/ZeebeClientBuilder.cs:135]: Ok nevermind I have re-read our PR description:

File: Client.UnitTests/CancelProcessInstanceTest.cs (hunk 1 above)
Comment (xlegalles) [reply Client.UnitTests/CancelProcessInstanceTest.cs:22]: I have changed the logic of these tests and share the same server across all. The execution is now way more faster. But I need to reset its state, otherwise the expectation is polluted by previous methods.

File: Client/Client.csproj (hunk 2 above)
Comment (xlegalles) [reply Client/Client.csproj:136]: The restoration process is able to manage transitive dependencies so that you don't need to specify them. If you do you will have to manage them, and ensure they are consistent with your primary dependencies.

File: Client/Impl/Builder/ZeebeClientBuilder.cs (hunk 3 above)
Comment (xlegalles) [reply Client/Impl/Builder/ZeebeClientBuilder.cs:135]: Yes this is the explanation.

File: Client/Impl/CustomHeaderInterceptor.cs (hunk 4 above)
Comment (xlegalles) [reply Client/Impl/CustomHeaderInterceptor.cs:11]: I wanted to maintain the behavior that you have implemented with a specific user-agent description. If I don't do this you will receive on the server, the generic description of a .net grpc agent.

**Comment by Zelldon** [comment]
Hey @xlegalles thank you for your time and effort!

I had a deeper look at this issue and it seemed to me that most of the changes were not really necessary. 

I made separate PRs to migrate to .net 7 #543 and to migrate to grpc-dotnet #545, both involved not much (or none) refactoring. 

I will implement/add some more features which have been recently added to Zeebe and then release a v2.0. version. 

Thanks again and sorry that I was not able to merge your PR. But I hope that you're happy with the end result anyway and it helps you in your project as well.

P.S.: 

Small recommendation, for the next time, please try to create smaller PRs (with fewer changes) it makes it easier to review and reason about. For example, try to split bigger changes up. Plus try to discuss the changes you want to do first in a separate issue, so you not working on changes unnecessarily which might not get accepted or merged.

**Comment by xlegalles** [comment]
Hi Christopher,
I understand your point of view and will do differently next time. And of
course, I'm satisfied that we can use the new version.
Thank you

Xavier
